### ISSUE TITLE
BlenderBIM 230115 doesn't load JSON results

### ISSUE THREAD

**Issue by imebp85** [IfcOpenShell/IfcOpenShell#2698]
Hi,

I have downloaded BlenderBIM from the website (230107) and have been stumbling upon this issue: https://github.com/IfcOpenShell/IfcOpenShell/issues/2675

The last comment recommended to download latest build (230115) to solve that issue.
And latests build actually solves it. But now I am facing some different issues:

1. When "Execute IFC clash", the tool saves the result as BCF, but "Select IFC results" expects a JSON. 
After some struggling, I have found that changing the extension of the file actually changes the export type, so I have been able to save the result as JSON just by changing the filename in the file dialog. Is this correct?
2. After "Execute IFC clash", nothing is selected in the model (in the project there is an actual clash, which is reported in the JSON file). Shouldn't it show the clash somehow?
3. With the JSON generated above, I use it in "Select IFC results", but now I am getting the following error:

![image](https://user-images.githubusercontent.com/122728940/212540196-34ca8cb0-4a1d-41a4-88a1-bb5e1dbfd936.png)

Am I doing something wrong or is this a known issue?

Thank you very much for your help,
Best regards,
Imelda

**Comment by sanzoghenzo** [comment]
Hi @imebp85 ,
you're right, the "Select IFC clash results" only supports json file as input as of now.
If I understand the code correctly, we need to implement a method that returns the global ids of the element in a clash set -> of an entire BCF file, since clash sets are exported separately.

As per your issue, it is indeed a bug, the operator misses a call to get the current IFC file.
You should be able to fix it adding the following line just before the line 290 of ...blenderbim\bim\module\clash\operator.py:

```
self.file = IfcStore.get_file()
```

Let me know if you try this and everything works fine, so I can commit the fix in a official way!

**Comment by sanzoghenzo** [comment]
@Moult : I still have troubles understanding the operators architecture, so I don't know how to change the `ExportClashSets` class to accept both bcf and json files in the dialog: should I add a `filter_glob` and drop the `filter_ext` attribute?

**Comment by Moult** [comment]
@sanzoghenzo that's correct.

**Comment by Moult** [comment]
The numbers looking the same is a coincidence, and this is a separate issue to what @sanzoghenzo was working on. Can you confirm that you only have a single IFC model loaded in your Blender session? Is that model exactly in sync with the IFC?

**Comment by imebp85** [comment]
Hi @Moult 
No. I am using two IFCs, one for architecture and structure, and one for MEP. 
I imported both IFCs in Blender, and wanted to try the Clash Detection tool.

My procedure is as follows:
- Import both files into blender.
- Add each IFC file as a Clash Source (architecture and structure in Group A, and MEP in Group B)
- Include `.IfcColumn` for Group A, and `.IfcPipeSegment` for Group B (which are the clashing object types). 

Note that the IFCs imported in Blender are the same ones I added as Clash Source.

Is this procedure correct or am I doing something wrong?

**Comment by Moult** [comment]
Right, so the first step is where things go wrong, don't "import both files", because right now we only support a single file per Blender session. You can have the second file in a separate model and link it back into the first model, but it will be a separate blend file. That should explain the issue.

**Comment by imebp85** [comment]
Hi @Moult 
I've saved both IFCs as .blender files, when I try to run IFC clash results in each model separately it works (in the mep model it selects the pipe, and in the architectural/structural model it selects the column).
So each individual model is working fine, as you said, the issue was importing both files.

Now I'm trying to link the MEP model into the Architectural/Structural model as you explained.
I've linked the Scenes, Collections, Objects... and I always have the following error when selecting the IFC clash results: 

```
Python: Traceback (most recent call last):
  File "C:\Users\imeld\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\addons\blenderbim\bim\module\clash\operator.py", line 292, in execute
    if element.GlobalId in global_ids:
  File "C:\Users\imeld\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\entity_instance.py", line 122, in __getattr__
    raise AttributeError(
AttributeError: entity instance of type 'IFC4.IfcSurfaceStyleRendering' has no attribute 'GlobalId'
```

I'm sure I'm not linking it correctly. Is there an example I can follow? or can you please provide some explanation about how to link it correctly? 

Thank you very much, 

Imelda

**Comment by Moult** [comment]
Yes, can you try linking with the Scene Properties > IFC Project Setup > IFC Links, press the "Link IFC" then (unintuitively) browse to the other .blend file.

**Comment by imebp85** [comment]
Hi @Moult 

I just liked the `.blend` file the way you told me, but when I press `Select IFC Clash Results` it results in the same error: 

```
Python: Traceback (most recent call last):
  File "C:\Users\imeld\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\addons\blenderbim\bim\module\clash\operator.py", line 292, in execute
    if element.GlobalId in global_ids:
  File "C:\Users\imeld\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\entity_instance.py", line 122, in __getattr__
    raise AttributeError(
AttributeError: entity instance of type 'IFC4.IfcSurfaceStyleRendering' has no attribute 'GlobalId'
```

May this be another issue with the library change, or am I doing something wrong? 

Thank you so much, 

Imelda

**Comment by imebp85** [comment]
Hi @Moult @sanzoghenzo 
I've been trying today with the new release and the issue above is still happenning: 

```
AttributeError: entity instance of type 'IFC4.IfcSurfaceStyleRendering' has no attribute 'GlobalId'
```

I have followed the linking process suggested by Moult.

I am a bit lost. Can you help me please?

Thank you, 

Best regards.

**Comment by Moult** [comment]
I've added it to the todo list but there is a bit of a backlog to get through.

**Comment by sanzoghenzo** [comment]
That's strange, my guess after seeing the error and looking at the code is that the Ifc IDs gets mixed up, and instead of getting a "physical" object derived by IfcRoot (that has the GlobalId attribute) it gets the "IfcSurfaceStyleRendering" that shouldn't even be considered in a clash test.
Could it be that the operator is using the ids of file "A" but tries to query file "B" with them?

**Comment by Moult** [comment]
Indeed it is a linking problem - right now the Blender session assumes only one active IFC in memory, not multiple, so it'll fetch random garbage data if you're linking from another Blender session.

**Comment by imebp85** [comment]
Hi @Moult and @sanzoghenzo
Thank you so much for your responses. Please let me know once the issue is solved. 
Best regards.

**Comment by imebp85** [comment]
Hi @Moult I've seen you closed the issue, but I have tried the BlenderBIM version 20230302, and I keep getting this issue: 

```
Python: Traceback (most recent call last):
  File "C:\Users\imeld\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\addons\blenderbim\bim\module\clash\operator.py", line 312, in execute
    if element.GlobalId in global_ids:
  File "C:\Users\imeld\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\entity_instance.py", line 173, in __getattr__
    raise AttributeError(
AttributeError: entity instance of type 'IFC4.IfcSurfaceStyleRendering' has no attribute 'GlobalId'

```
Am I doing something wrong or the issue still exists? 

Thank you, best regards.

**Comment by Moult** [comment]
That looks like an unrelated issue, how are you loading your IFCs?

**Comment by imebp85** [comment]
Hi @Moult 

- I imported each ifc file (MEP and Architecture) and saved each one as a .blend file. 
-  I opened the architecture.blend file.
-  I linked the MEP file with the Scene Properties > IFC Project Setup > IFC Links, press the "Link IFC" then (unintuitively) browse to the other architecture.blend file.
- In Scene Properties > IFC Quality Control > IFC Clash Sets, I added a new clash set to check the architecture.ifc in Group A, and MEP.ifc file in Group B
-  Then included .IfcColumn for Group A, and .IfcDuctSegment for Group B (which are the clashing object types).
-  After that I generated the json with the "Execute IFC Clash" button. 
- Then when I load the json results with the button "Select IFC Clash Results" it is when that error pops up. The column gets highlighted, but the duct segment doesn't. 

I have attached the files here in case you want to try them. 

Thank you very much, 

Imelda 
[Casa Perro v3.0.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/10900877/Casa.Perro.v3.0.zip)
